### DISCUSSION TITLE
Sending multiple signals in a single harchardware invocation

### DISCUSSION THREAD

**Discussion by JanCeuleers** [bengtmartensson/HarcHardware#31] [General]
I am preparing to use harchardware to control a set-top box. It uses the TDC-38 protocol, and I can change to channel 121 as follows:

$ harchardware --arduino --device /dev/ttyUSB0 transmit --protocol TDC-38 --names D=18,F=2,S=10
$ harchardware --arduino --device /dev/ttyUSB0 transmit --protocol TDC-38 --names D=18,F=5,S=10
$ harchardware --arduino --device /dev/ttyUSB0 transmit --protocol TDC-38 --names D=18,F=2,S=10

However, this takes quite a long time: upon each invocation the JVM has to start, so that I'm concerned that this method may not be robust in view of the STB's inter-digit timeout.

I was therefore wondering whether it is possible to send multiple signals in one invocation. I have already tried repeating the --protocol and --names options on the command line but that doesn't work.

**Comment by JanCeuleers** [comment]
Update: the issue with inter-digit timeout has materialised; I really do need a way to speed the process up.

**Comment by bengtmartensson** [comment]
First of all, I can confirm that you have understood things correctly; there is nothing obvious that you have missed. HarcHardware is simply not intended for your use case. (A project that does address that usecase is [JGirs](https://github.com/bengtmartensson/JGirs), but, although "finished" in some sense, is somewhat ... rough...). One sledge-hammer approach would be to manually piece together raw signals to implement the desired sequence. If Java programming belongs to your skillset, you can also write a small program (should be no more than one page at most) that sends such sequences.

Having said that, we actually do not have anything (except JGirs...) for sending such sequences. Usage would be, not only to turn TVs to multi-digit channels, but also sequences like sending misc. "magic sequences", for example for region-unlocking DVD-player (the Panasonic F2 sequence, for example). I may come up with something. But it will hardly be anything that helps you in the short time span.

**Comment by JanCeuleers** [comment]
Thanks for your reply, and understood. I don't mind doing a bit of programming myself, but Java isn't among my skillset I'm afraid.

What I now have in mind is your suggestion on determining the raw signals and sending those straight to the Arduino (i.e. bypassing harchardware altogether.

The question then becomes how to capture the GIRS commands that harchardware or irscrutinizer issue to the Arduino. I have tried the following without success:

$ harchardware --arduino --device /dev/pts/5 send 1 38000 0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 315 315 630 630 315 315 630 89315

in which the "send 1 38000..." was scraped from irscrutinizer's logging.

I have also tried using interceptty to sniff serial port traffic, but harchardware refuses to connect to the pty:

[org.harctoolbox.harchardware.Main processCommand] SEVERE: No such port: /dev/pts/5

although it does exist, and I can use minicom to interact with the arduino through interceptty by means of that pty.

**Reply by bengtmartensson** [reply]
Try the -v (--verbose) command line option, in IrScrutinizer menu item Options -> verbose.

**Comment by JanCeuleers** [comment]
Got it, thank you.

I wrote a bash script that generates an expect script, thusly:

```
#!/bin/bash
# $1: transmitter
# $2: channel number (1 or more digits)

case $1 in
  1)
    export device=/dev/ttyUSB0
    ;;
  2)
    export device=/dev/ttyUSB1
    ;;
  3)
    export device=/dev/ttyUSB2
    ;;
  *)
    exit 1
    ;;
esac

export numstr=$2
cmds=" "
filename=/tmp/expect$1

cat << EOF > ${filename}
#!/usr/bin/expect -f
set portID [open ${device} r+]
set baud 115200
fconfigure \$portID -mode "115200,n,8,1"
fconfigure \$portID -blocking 0 -buffering none

spawn -open \$portID
set timeout 2
send -- "\r"
expect "OK\r\n"
EOF
for (( i=0; i<${#numstr}; i++ )); do
  cmd=`fgrep ${numstr:$i:1}: sendcmds.txt | awk '{$1=""; print $0}'`
  echo send -- \"send 1 38000 ${cmd}\\r\" >> ${filename}
  echo expect \"OK\\r\\n\" >> ${filename}
done

/usr/bin/expect -f ${filename}
```
The parameters of the send commands are read from a file, which I manually constructed from verbose irscrutinizer output (omitting the "send 1 38000 " from the beginning of each command):

Power:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 630 630 630 315 315 630 315 89000
1:	0 28 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 315 315 315 315 315 315 630 630 315 89000
2:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 315 315 315 315 630 630 630 89315
3:	0 28 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 315 315 315 315 630 315 315 630 315 89000
4:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 315 315 630 630 315 315 630 89315
5:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 315 315 630 630 630 630 315 89000
6:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 315 315 630 315 315 630 630 89315
7:	0 28 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 315 315 630 315 315 315 315 630 315 89000
8:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 630 630 315 315 315 315 630 89315
9:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 630 630 315 315 630 630 315 89000
0:	0 24 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 315 315 630 630 630 630 630 89315
Red:	0 24 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 630 630 315 315 630 315 315 630 630 89315
TXT:	0 24 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 630 630 315 315 630 630 630 630 315 89000
Del:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 315 315 315 315 630 315 315 630 89315
Home:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 630 315 315 315 315 315 315 630 89315
OK:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 315 315 630 630 315 315 630 315 89000
Up:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 630 630 315 315 315 315 630 315 89000
Down:	0 24 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 315 315 630 630 630 630 89315
Left:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 315 315 630 315 315 315 315 630 89315
Right:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 315 315 630 315 315 630 630 315 89000
Back:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 315 315 315 315 315 315 630 630 89315
Exit:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 315 315 315 315 315 315 630 630 89315
VolUp:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 630 315 315 630 315 315 630 315 89000
VolDown:	0 24 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 630 630 630 315 315 630 89315
ChanUp:	0 24 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 630 630 630 630 630 315 89000
ChanDown:	0 24 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 630 630 315 315 630 630 89315
Mute:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 630 315 315 315 315 630 630 315 89000
Stop:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 630 630 315 315 315 315 630 315 315 630 315 89000
Rec:	0 24 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 630 630 315 315 315 315 630 630 630 89315
Rewind:	0 28 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 315 315 630 315 315 315 315 315 315 315 315 630 315 89000
FForward:	0 26 0 315 315 315 630 315 315 630 630 315 315 630 630 630 630 630 630 315 315 315 315 315 315 315 315 630 89315

Of course this is purely a channel changing script; it can't send signals other than digits, but that's easy to add if necessary.

**Comment by bengtmartensson** [comment]
I have implemented a (very) simple macro facility, see #32. I think this will fulfill your requirements as listed here. The  stuff in the original posting can be made as

    $ harchardware --arduino --device /dev/ttyUSB0 transmit --file arguments.txt

where the file `arguments.txt` looks as follows:

    --protocol TDC-38 --names D=18,F=2,S=10
    --protocol TDC-38 --names D=18,F=5,S=10
    --protocol TDC-38 --names D=18,F=2,S=10

The lines are simply what is to the right of `transmit` from the original command line. Whaddyathink?

Do you know `/dev/serial/by-id`?

**Comment by JanCeuleers** [comment]
Hi Bengt,

Thank you for looking into this further.

That would work, and I will give it a try, but I've already swung my
expect-based method into production. Very fast -- no JVM to load.

My Arduinos are identical and do not report a serial number, so I have
had to teach udev to assign them names based on the physical ports they
are plugged into.

Cheers, Jan
